### PR TITLE
MSYS build scripts and instructions

### PR DESCRIPTION
**Pull request by Stinkfist0**
Opening up for an initial review; a commit or two probably pending.

### CODE REVIEW

**Comment by iras** [comment]
Thanks @Stinkfist0.

**Comment by Stinkfist0** [comment]
Closes #98 also

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -490,10 +492,14 @@ endif()
 get_filename_component(libTKernelExt ${libTKernel} EXT)
 if("${libTKernelExt}" STREQUAL ".a")
     find_package(Threads)
-    if (NOT APPLE)
-        set(LIB_RT "rt")
+    # TODO why OPENCASCADE_LIBRARIES is repeated three times below?
```
Comment (aothms) [inline cmake/CMakeLists.txt]: This is to resolve cyclic dependencies in gcc/ld. Better way seems to use `--start-group ... --end-group`, but didn't want to spend effort on it.

**Comment by aothms** [comment]
This is great thanks. Are there more commits pending? Don't really have the opportunity to test the changes, but they all look perfectly reasonable and since the bulk comprises build script additions, I am fine with merging as is.

**Comment by Stinkfist0** [comment]
I'm having missing symbols errors for Python (Python is found though - any tips?), but other than that this should be good to go.

**Comment by iras** [comment]
Most of the warnings have gone when compiling in gcc. As @Stinkfist0 already pointed out there are some warnings re Python.

`Scanning dependencies of target IfcParse
[  2%] [  5%] [  8%] [ 13%] [ 13%] [ 16%] [ 19%] Building CXX object CMakeFiles/IfcParse.dir/home/i/git/IfcOpenShell-2/src/ifcparse/IfcLateBoundEntity.cpp.o
[ 22%] Building CXX object CMakeFiles/IfcParse.dir/home/i/git/IfcOpenShell-2/src/ifcparse/IfcParse.cpp.o
Building CXX object CMakeFiles/IfcParse.dir/home/i/git/IfcOpenShell-2/src/ifcparse/IfcSIPrefix.cpp.o
Building CXX object CMakeFiles/IfcParse.dir/home/i/git/IfcOpenShell-2/src/ifcparse/IfcCharacterDecoder.cpp.o
Building CXX object CMakeFiles/IfcParse.dir/home/i/git/IfcOpenShell-2/src/ifcparse/IfcLogger.cpp.o
Building CXX object CMakeFiles/IfcParse.dir/home/i/git/IfcOpenShell-2/src/ifcparse/IfcUtil.cpp.o
Building CXX object CMakeFiles/IfcParse.dir/home/i/git/IfcOpenShell-2/src/ifcparse/Ifc2x3-latebound.cpp.o
Building CXX object CMakeFiles/IfcParse.dir/home/i/git/IfcOpenShell-2/src/ifcparse/IfcGlobalId.cpp.o
[ 25%] Building CXX object CMakeFiles/IfcParse.dir/home/i/git/IfcOpenShell-2/src/ifcparse/IfcSpfHeader.cpp.o
[ 27%] Building CXX object CMakeFiles/IfcParse.dir/home/i/git/IfcOpenShell-2/src/ifcparse/IfcHierarchyHelper.cpp.o
/home/i/git/IfcOpenShell-2/src/ifcparse/Ifc2x3-latebound.cpp: In function ‘void InitDescriptorMap()’:
/home/i/git/IfcOpenShell-2/src/ifcparse/Ifc2x3-latebound.cpp:2426:31: warning: variable ‘current_enum’ set but not used [-Wunused-but-set-variable]
     IfcEnumerationDescriptor* current_enum;
                               ^
[ 30%] Building CXX object CMakeFiles/IfcParse.dir/home/i/git/IfcOpenShell-2/src/ifcparse/IfcWrite.cpp.o
[ 33%] Building CXX object CMakeFiles/IfcParse.dir/home/i/git/IfcOpenShell-2/src/ifcparse/Ifc2x3.cpp.o
/home/i/git/IfcOpenShell-2/src/ifcparse/IfcParse.cpp: In member function ‘IfcUtil::IfcBaseClass* IfcParse::IfcFile::addEntity(IfcUtil::IfcBaseClass*)’:
/home/i/git/IfcOpenShell-2/src/ifcparse/IfcParse.cpp:1255:34: warning: ‘*((void*)& conversion_factor +8)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
       (*it) *= *conversion_factor;
                                  ^
Linking CXX static library libIfcParse.a
[ 33%] Built target IfcParse
Scanning dependencies of target IfcGeom
Scanning dependencies of target IfcParseExamples
[ 36%] Building CXX object examples/CMakeFiles/IfcParseExamples.dir/IfcParseExamples.cpp.o
[ 38%] [ 41%] Building CXX object CMakeFiles/IfcGeom.dir/home/i/git/IfcOpenShell-2/src/ifcgeom/IfcGeomHelpers.cpp.o
[ 44%] Building CXX object CMakeFiles/IfcGeom.dir/home/i/git/IfcOpenShell-2/src/ifcgeom/IfcGeomFaces.cpp.o
Building CXX object CMakeFiles/IfcGeom.dir/home/i/git/IfcOpenShell-2/src/ifcgeom/IfcGeomShapes.cpp.o
[ 47%] [ 50%] Building CXX object CMakeFiles/IfcGeom.dir/home/i/git/IfcOpenShell-2/src/ifcgeom/IfcRegister.cpp.o
Building CXX object CMakeFiles/IfcGeom.dir/home/i/git/IfcOpenShell-2/src/ifcgeom/IfcGeomRepresentation.cpp.o
[ 55%] [ 52%] Building CXX object CMakeFiles/IfcGeom.dir/home/i/git/IfcOpenShell-2/src/ifcgeom/IfcGeomSerialisation.cpp.o
Building CXX object CMakeFiles/IfcGeom.dir/home/i/git/IfcOpenShell-2/src/ifcgeom/IfcGeomFunctions.cpp.o
Linking CXX executable IfcParseExamples
[ 58%] Building CXX object CMakeFiles/IfcGeom.dir/home/i/git/IfcOpenShell-2/src/ifcgeom/IfcGeomCurves.cpp.o
[ 58%] Built target IfcParseExamples
[ 61%] Building CXX object CMakeFiles/IfcGeom.dir/home/i/git/IfcOpenShell-2/src/ifcgeom/IfcGeomMaterial.cpp.o
[ 63%] Building CXX object CMakeFiles/IfcGeom.dir/home/i/git/IfcOpenShell-2/src/ifcgeom/IfcGeomWires.cpp.o
[ 66%] Building CXX object CMakeFiles/IfcGeom.dir/home/i/git/IfcOpenShell-2/src/ifcgeom/IfcGeomRenderStyles.cpp.o
Linking CXX static library libIfcGeom.a
[ 66%] Built target IfcGeom
[ 69%] Scanning dependencies of target IfcGeomServer
Scanning dependencies of target IfcAdvancedHouse
Scanning dependencies of target IfcOpenHouse
Scanning dependencies of target IfcConvert
Swig source
[ 72%] Building CXX object examples/CMakeFiles/IfcAdvancedHouse.dir/IfcAdvancedHouse.cpp.o
[ 75%] Building CXX object CMakeFiles/IfcGeomServer.dir/home/i/git/IfcOpenShell-2/src/ifcgeomserver/IfcGeomServer.cpp.o
[ 77%] Building CXX object examples/CMakeFiles/IfcOpenHouse.dir/IfcOpenHouse.cpp.o
[ 80%] [ 83%] [ 86%] [ 88%] Building CXX object CMakeFiles/IfcConvert.dir/home/i/git/IfcOpenShell-2/src/ifcconvert/OpenCascadeBasedSerializer.cpp.o
Building CXX object CMakeFiles/IfcConvert.dir/home/i/git/IfcOpenShell-2/src/ifcconvert/IfcConvert.cpp.o
Building CXX object CMakeFiles/IfcConvert.dir/home/i/git/IfcOpenShell-2/src/ifcconvert/ColladaSerializer.cpp.o
Building CXX object CMakeFiles/IfcConvert.dir/home/i/git/IfcOpenShell-2/src/ifcconvert/SvgSerializer.cpp.o
[ 91%] Building CXX object CMakeFiles/IfcConvert.dir/home/i/git/IfcOpenShell-2/src/ifcconvert/util.cpp.o
Scanning dependencies of target _ifcopenshell_wrapper
[ 91%] Swig source
[ 94%] Building CXX object CMakeFiles/IfcConvert.dir/home/i/git/IfcOpenShell-2/src/ifcconvert/XmlSerializer.cpp.o
[ 97%] Building CXX object ifcwrap/CMakeFiles/_ifcopenshell_wrapper.dir/IfcPythonPYTHON_wrap.cxx.o
/home/i/git/IfcOpenShell-2/src/ifcgeomserver/IfcGeomServer.cpp: In instantiation of ‘T sread(std::istream&) [with T = int; std::istream = std::basic_istream<char>]’:
/home/i/git/IfcOpenShell-2/src/ifcgeomserver/IfcGeomServer.cpp:58:32:   required from here
/home/i/git/IfcOpenShell-2/src/ifcgeomserver/IfcGeomServer.cpp:53:18: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  return *((T*)buf);
                  ^
/home/i/git/IfcOpenShell-2/src/ifcgeomserver/IfcGeomServer.cpp: In instantiation of ‘T sread(std::istream&) [with T = double; std::istream = std::basic_istream<char>]’:
/home/i/git/IfcOpenShell-2/src/ifcgeomserver/IfcGeomServer.cpp:299:32:   required from here
/home/i/git/IfcOpenShell-2/src/ifcgeomserver/IfcGeomServer.cpp:53:18: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
/home/i/git/IfcOpenShell-2/src/ifcgeomserver/IfcGeomServer.cpp: In instantiation of ‘T sread(std::istream&) [with T = unsigned int; std::istream = std::basic_istream<char>]’:
/home/i/git/IfcOpenShell-2/src/ifcgeomserver/IfcGeomServer.cpp:315:26:   required from here
/home/i/git/IfcOpenShell-2/src/ifcgeomserver/IfcGeomServer.cpp:53:18: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
[100%] Building CXX object CMakeFiles/IfcConvert.dir/home/i/git/IfcOpenShell-2/src/ifcconvert/WavefrontObjSerializer.cpp.o
Linking CXX executable IfcAdvancedHouse
[100%] Built target IfcAdvancedHouse
Linking CXX executable IfcOpenHouse
[100%] Built target IfcOpenHouse
Linking CXX executable IfcGeomServer
[100%] Built target IfcGeomServer
/home/i/git/IfcOpenShell-2/src/ifcconvert/IfcConvert.cpp: In function ‘int main(int, char**)’:
/home/i/git/IfcOpenShell-2/src/ifcconvert/IfcConvert.cpp:407:108: warning: ‘*((void*)& bounding_height +4)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
             static_cast<SvgSerializer*>(serializer)->setBoundingRectangle(*bounding_width, *bounding_height);
                                                                                                            ^
/home/i/git/IfcOpenShell-2/src/ifcconvert/IfcConvert.cpp:407:108: warning: ‘*((void*)& bounding_width +4)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
Linking CXX executable IfcConvert
[100%] Built target IfcConvert
Linking CXX shared module _ifcopenshell_wrapper.so
[100%] Built target _ifcopenshell_wrapper
`

**Comment by Stinkfist0** [comment]
Yeah couple warnings left, I could maybe take a look of them now.

**Comment by Stinkfist0** [comment]
Only GCC warning left (with the current warning level) should be `dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]` in IfcGeomServer.cpp, but I'll leave that for Thomas, as I don't have any experience in running and testing IfcGeomServer myself.

**Comment by iras** [comment]
Cool. Thank you @Stinkfist0. I recompiled it with gcc and I can confirm the only warning left is the one thrown by IfcGeomServer.cpp.

`[ 86%] Building CXX object CMakeFiles/IfcGeomServer.dir/home/i/git/IfcTest/src/ifcgeomserver/IfcGeomServer.cpp.o
/home/i/git/IfcTest/src/ifcgeomserver/IfcGeomServer.cpp: In instantiation of ‘T sread(std::istream&) [with T = int; std::istream = std::basic_istream<char>]’:
/home/i/git/IfcTest/src/ifcgeomserver/IfcGeomServer.cpp:58:32:   required from here
/home/i/git/IfcTest/src/ifcgeomserver/IfcGeomServer.cpp:53:18: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  return *((T*)buf);
                  ^
/home/i/git/IfcTest/src/ifcgeomserver/IfcGeomServer.cpp: In instantiation of ‘T sread(std::istream&) [with T = double; std::istream = std::basic_istream<char>]’:
/home/i/git/IfcTest/src/ifcgeomserver/IfcGeomServer.cpp:299:32:   required from here
/home/i/git/IfcTest/src/ifcgeomserver/IfcGeomServer.cpp:53:18: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
/home/i/git/IfcTest/src/ifcgeomserver/IfcGeomServer.cpp: In instantiation of ‘T sread(std::istream&) [with T = unsigned int; std::istream = std::basic_istream<char>]’:
/home/i/git/IfcTest/src/ifcgeomserver/IfcGeomServer.cpp:315:26:   required from here
/home/i/git/IfcTest/src/ifcgeomserver/IfcGeomServer.cpp:53:18: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
`

Thumbs up from me if you want to merge it @aothms.

**Comment by aothms** [comment]
Wonderful. I am still travelling, no time to check in detail, but looks good to me. Many many thanks.

File: cmake/CMakeLists.txt [hunk 2]
```diff
@@ -291,8 +291,9 @@ if(ENABLE_BUILD_OPTIMIZATIONS)
 		set(CMAKE_EXE_LINKER_FLAGS_RELEASE "${CMAKE_EXE_LINKER_FLAGS_RELEASE} /LTCG /OPT:REF")
 		set(CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO "${CMAKE_EXE_LINKER_FLAGS_RELEASE} /DEBUG /OPT:NOICF")
 	else()
-		#TODO GCC (& Clang?) optimizations
-		message(STATUS "ENABLE_BUILD_OPTIMIZATIONS not implemented for GCC/non-MSVC compilers.)")
+        # GCC-like: Release should use O3 but RelWithDebInfo 02 so enforce 03. Anything other useful that could be added here?
+        set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3")
+        set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELEASE} -O3")
```
Comment (Stinkfist0) [inline cmake/CMakeLists.txt]: Could also experiment with `-flto` and `-fwhole-program` here.

**Pull request merged**